util/oci/client.go

```diff
@@ -0,0 +1,544 @@
+func (c *nativeOCIClient) CleanCache(revision, project string) error {
+	cachePath, err := c.getCachedPath(revision, project)
+	if err != nil {
+		return err
+	}
+	return os.RemoveAll(cachePath)
+}
+
+// DigestMetadata extracts the OCI manifest for a given revision and returns it to the caller.
+func (c *nativeOCIClient) DigestMetadata(ctx context.Context, digest, project string) (*v1.Manifest, error) {
+	path, err := c.getCachedPath(digest, project)
+	if err != nil {
+		return nil, err
+	}
+
+	repo, err := oci.NewFromTar(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+
+	return getOCIManifest(ctx, digest, repo)
+}
+
+func (c *nativeOCIClient) ResolveRevision(ctx context.Context, revision string, noCache bool) (string, error) {
+	constraints, err := semver.NewConstraint(revision)
+	if err == nil {
+		tags, err := c.getTags(ctx, noCache)
+		if err != nil {
+			return "", fmt.Errorf("error fetching tags: %w", err)
+		}
+		version, err := tags.MaxVersion(constraints)
+		if err != nil {
+			return "", fmt.Errorf("no version for constraints: %w", err)
+		}
+		return c.resolveDigest(ctx, version.String())
+	}
+
+	return c.resolveDigest(ctx, revision)
+}
+
+func (c *nativeOCIClient) getTags(ctx context.Context, noCache bool) (*TagsList, error) {
+	indexLock.Lock(c.repoURL)
+	defer indexLock.Unlock(c.repoURL)
+
+	var data []byte
+	if !noCache && c.indexCache != nil {
+		if err := c.indexCache.GetHelmIndex(c.repoURL, &data); err != nil && !errors.Is(err, cache.ErrCacheMiss) {
+			log.Warnf("Failed to load index cache for repo: %s: %s", c.repoLock, err)
+		}
+	}
+
+	tags := &TagsList{}
+	if len(data) == 0 {
+		start := time.Now()
+		result, err := c.tagsFunc(ctx, "")
+		if err != nil {
+			return nil, fmt.Errorf("failed to get tags: %w", err)
+		}
+
+		for _, tag := range result {
+			// By convention: Change underscore (_) back to plus (+) to get valid SemVer
+			convertedTag := strings.ReplaceAll(tag, "_", "+")
+			tags.Tags = append(tags.Tags, convertedTag)
+		}
+
+		log.WithFields(
+			log.Fields{"seconds": time.Since(start).Seconds(), "repo": c.repoURL},
+		).Info("took to get tags")
+
+		if c.indexCache != nil {
+			if err := c.indexCache.SetHelmIndex(c.repoURL, data); err != nil {
+				log.Warnf("Failed to store tags list cache for repo: %s: %s", c.repoURL, err)
+			}
+		}
+	} else {
+		err := json.Unmarshal(data, tags)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode tags: %w", err)
+		}
+	}
+
+	return tags, nil
+}
+
+// resolveDigest resolves a digest from a tag.
+func (c *nativeOCIClient) resolveDigest(ctx context.Context, revision string) (string, error) {
+	descriptor, err := c.repo.Resolve(ctx, revision)
+	if err != nil {
+		return "", fmt.Errorf("cannot get digest for revision %s: %w", revision, err)
+	}
+
+	return descriptor.Digest.String(), nil
+}
+
+func newTLSConfig(creds Creds) (*tls.Config, error) {
```

Comment:
Is this code duplicated from other client(s)? Should it be abstracted?

Reply:
Yeah this was originally duplicated from the Helm implementation. I'll see what can be done to abstract it out, but this is already a pretty big PR. Perhaps this abstraction can be done in a separate PR, and then I can rebase off it when it's merged?

The same thing applies for the semver implementation, there's an existing PR #20216 which I can update to if it gets merged, so I can make use of that implementation instead.